### ISSUE TITLE
Using Expressions for data binding

### ISSUE THREAD

**Issue by powerdude** [FmgLib/FmgLib.MauiMarkup#162]
We currently have to do bindings using strings like this:

```csharp
new Button().BackgroundColor(e => e.Path($"{nameof(SomeChildOfVM)}.{nameof(SomeChildOfVM.TestColor)}"))
```

It would be nice if an expression could be used instead to provide a more ergonomic experience and reach parity with other markup libraries.  So, instead of above, one could do this:

```csharp
new Button().BackgroundColor( e=> e.Path( static (ParentVm vm) => vm.SomeChildOfVM.TestColor));
```

While the above is a pattern followed in other libraries, there could be a way to remove some of the duplication that would show up if there is more than one property that needs to be bound.  Something like one of the options below could help:

```csharp
new Button().With<ParentVm>().BackgroundColor(e=>e.Path(x=>x.SomeChildOfVM.TestColor));
```
or

```csharp
new Button().BackgroundColor<ParentVm>(e=>e.Path(x=>x.SomeChildOfVM.TestColor));
```

The goal being to not have to specify the VM type too many times.

**Comment by gonultasmf** [comment]
have you ever tried using it like this:

```csharp
public partial class MainPage : ContentPage
{
	public MainPage(SomeChildOfVM viewModel)
	{
	     BindingContext = viewModel;
	     Build();
	}

	public void Build()
	{
		this
		.Content(
			new Button().BackgroundColor(e=>e.Path(nameof(BindingContext.TestColor)))
		);
	}
}
```
or
```csharp
public partial class MainPage : ContentPage
{
	private readonly SomeChildOfVM_1 _viewModel1;
	private readonly SomeChildOfVM_2 _viewModel2;

	public MainPage(SomeChildOfVM_1 viewModel1, SomeChildOfVM_2 viewModel2)
	{
		_viewModel1 = viewModel1;
		_viewModel2 = viewModel2;
		Build();
	}

	public void Build()
	{
		this
		.Content(
			new Button()
			.BackgroundColor(e=>e.Path(nameof(_viewModel1.TestColor)).Source(_viewModel1))
			.TextColor(e=>e.Path(nameof(_viewModel2.TestColor)).Source(_viewModel2))
		);
	}
}
```

Did you know how to use the Source method?
and does this usage work for you?

**Comment by powerdude** [comment]
Yes, I've tried the first, but not the second.  The challenges come with having complex VM.  Something like:

```csharp
public class Person
{
   public string FirstName {get;set;}
  public string LastName {get;set;}
}

public class Employee
{
     public PersonDetails Person {get;set;}
public string Title {get;set;}
}

public class EmployeeVM
{
   public Employee Employee {get;set;}
}

```

Now, usually you'd only inject the `EmployeeVM` into a page.  

So, now, create a label to show the employee's name.

```csharp
new Label().Text( e => e.Path("Employee.Person.FirstName");
```

It needs to be flexible enough to support binding in DataTemplates for things like `CollectionView` also.  

Does that help?

**Comment by gonultasmf** [comment]
Yes,
with this usage `(e => e.Path("Employee.Person.FirstName"))`
you will get the FirstName value in the Label Text property.


This feature has been tested in many applications for CollectionView. Let me leave an example usage here:
```csharp
public partial class CategoriesPage : BasePage<CategoriesPageViewModel>
{
    public CategoriesPage(CategoriesPageViewModel viewModel) : base(viewModel, "Kategoriler")
    {
    }

    public override void Build()
    {
        this
        .Content(
            new VerticalStackLayout()
            .Margin(10)
            .FillHorizontal()
            .Children(
                new CollectionView()
                .SelectionMode(SelectionMode.None)
                //.ItemsSource(e => e.Path("Categories"))
                .Bind(CollectionView.ItemsSourceProperty, "Categories")
                .ItemsLayout(
                    new LinearItemsLayout(ItemsLayoutOrientation.Vertical)
                    .ItemSpacing(10)
                )
                .EmptyView(
                    new VerticalStackLayout()
                    .Center()
                    .Children(
                        new Label()
                            .Text("Kayıt Yoktur.")
                            .TextColor(Colors.Red)
                            .FontAttributes(FontAttributes.Bold)
                            .FontSize(18)
                    )
                )
                .ItemTemplate(
                    () => 
                        new Expander()
                        .IsExpanded(true)
                        .Assign(out var expander)
                        .Header(
                            new Frame()
                            .BackgroundColor(Colors.Transparent)
                            .BorderColor(Colors.Transparent)
                            .Padding(0)
                            .Content(
                                new Grid()
                                .ColumnDefinitions(e => e.Star(9).Star(1))
                                .Margin(5)
                                .Children(
                                    new Label()
                                    .Text(e => e.Path(nameof(CategoryVM.Name)))
                                    //.Bind(Label.TextProperty, nameof(CategoryVM.Name))
                                    .FontSize(18)
                                    .FontAttributes(FontAttributes.Bold)
                                    .Column(0),

                                    new Image()
                                    .Source("chevron.png")
                                    .SizeRequest(25,25)
                                    .Column(1)
                                    .AlignBottomRight()
                                    .Assign(out var image)
                                    .Triggers(
                                        new MultiTrigger(typeof(Image))
                                        .Conditions(
                                            new BindingCondition()
                                            .Binding(b => b
                                                           .Source(expander)
                                                           .Path("IsExpanded"))
                                            .Value(true)
                                        )
                                        .Setters(new Setters<Image>(e => e.Rotation(180)))
                                    )
                                )
                            )
                        )
                        .Content(
                            new CollectionView()
                            .SelectionMode(SelectionMode.None)
                            .ItemsSource(e => e.Path(nameof(CategoryVM.SubCategories)))
                            //.Bind(CollectionView.ItemsSourceProperty, nameof(CategoryVM.SubCategories))
                            .ItemsLayout(
                                new GridItemsLayout(ItemsLayoutOrientation.Vertical)
                                .Span(2)
                                .HorizontalItemSpacing(10)
                                .VerticalItemSpacing(10)
                            )
                            .EmptyView(
                                new VerticalStackLayout()
                                .Center()
                                .Children(
                                    new Label()
                                        .Text("Kayıt Yoktur.")
                                        .TextColor(Colors.Red)
                                        .FontAttributes(FontAttributes.Bold)
                                        .FontSize(18)
                                )
                            )
                            .ItemTemplate(
                                () => new Frame()
                                        .CornerRadius(15)
                                        .BorderColor(Colors.LightGray)
                                        .BackgroundColor(Colors.LightGray)
                                        .MinimumHeightRequest(30)
                                        .WidthRequest(180)
                                        .Padding(0)
                                        .Margin(new Thickness(1, 0, 5, 5))
                                        .FlexBasis(FlexBasis.Auto)
                                        .Content(
                                            new Grid()
                                            .ColumnDefinitions(e => e.Star(3).Star(7))
                                            .Padding(5)
                                            .Children(
                                                new Image()
                                                .Source(e => e.Path(nameof(SubCategoryVM.Icon)))
                                                //.Bind(Image.SourceProperty, nameof(SubCategoryVM.Icon))
                                                .SizeRequest(30, 30)
                                                .Column(0)
                                                .CenterVertical(),

                                                new Label()
                                                .Text(e => e.Path(nameof(SubCategoryVM.Name)))
                                                //.Bind(Label.TextProperty, nameof(SubCategoryVM.Name))
                                                .TextColor(Colors.CornflowerBlue)
                                                .FontAttributes(FontAttributes.Bold)
                                                .FontSize(12)
                                                .Column(1)
                                                .FontAutoScalingEnabled(true)
                                                .CenterVertical()
                                            )
                                        )
                            )
                        )
                )
            )
        );
    }
}
```

**Comment by gonultasmf** [comment]
I am also leaving a repo of a sample application made with DevExpress if you want to check it out: https://github.com/gonultasmf/MyFinance

**Comment by gonultasmf** [comment]
support for this problem, similar to the one in the Maui toolkit, will be added to FmgLib.MauiMarkup.

My goal is to achieve a more developer-friendly structure.

**Comment by powerdude** [comment]
the hoped goal is to not have to do `Employee.Person.FirstName` as that won't survive refactoring very well.  Even `{nameof(EmployeeVM.Employee)}.{nameof(Employee.Person)}.{nameof(Person.FirstName)}` isn't great because if this structure changes, there will be no compile time errors and things will only be found at run-time.  It is one of the main reasons to use expressions and results in less verbose-ness and mental gymnastics with creating the string.

**Comment by gonultasmf** [comment]
Released with version 8.8.9

**Comment by gonultasmf** [comment]
@powerdude With new version 9.0.2, there is now a usage like this:

```csharp
new Label().Text(e => e.Getter(static (MyViewModel model) => model.Username));
```
